sdcm/utils/uda.py

```diff
@@ -0,0 +1,109 @@
+        RETURNS tuple<bigint,bigint>
+        LANGUAGE lua as 'return {acc[1] + 1, acc[2] + #a}';
+
+        CREATE OR REPLACE FUNCTION present(res tuple<bigint,bigint>)
+        RETURNS NULL ON NULL INPUT
+        RETURNS text
+        LANGUAGE lua as 'return "The average string length is " .. res[2]/res[1] .. "!"';
+
+        CREATE OR REPLACE AGGREGATE avg_length(text)
+        SFUNC accumulate_len
+        STYPE tuple<bigint,bigint>
+        FINALFUNC present INITCOND (0,0);
+
+    (Optionally you can also add a REDUCEFUNC <reduce_func_name> if running a
+    map-reduce style distributed workload.)
+    UDAs are bound to keyspaces, i.e. they cannot be used outside of the keyspace they were
+    created in.
+
+    Using the custom UDA is similar to usual CQL functions. Assuming we're using the keyspace
+    the UDA was created in:
+
+        SELECT avg_length(word) FROM words;
+
+    The class instance can be created explicitly by providing the necessary attributes (it provides some
+    baseline type validation by subclassing PyDantic's BaseModel).
+    The preferred way to create UDAs is by creating a yaml file with the same fields as this class's attributes
+    and getting the UDA instance by using the from_yaml() class method. The fields need to specify names of
+    existing UDFs (i.e. UDFs that can be loaded before the UDA is loaded).
+
+    UDF/UDA presentation from 2019 summit: https://www.scylladb.com/tech-talk/udf-uda-and-whats-in-the-future/
+    UDF/UDA testing desing doc: https://docs.google.com/document/d/16GTe1bLmMBC5IVCjC_nY-UNnMCLr_3V2C6K6tiYPstQ
+    /edit?usp=sharing
+    """
+    name: str
+    args: str
+    return_type: str
+    accumulator_udf: UDF
+    reduce_udf: Optional[UDF]
+    final_udf: UDF
+    initial_condition: str
+
+    def get_create_query_string(self, ks: str):
+        query_string = f"CREATE AGGREGATE {ks}.{self.name}" \
+                       f"({self.args}) SFUNC {self.accumulator_udf.name} " \
+                       f'STYPE {self.accumulator_udf.return_type} '
+        if self.reduce_udf:
+            query_string += f"REDUCEFUNC {self.reduce_udf.name} "
+
+        query_string += f"FINALFUNC {self.final_udf.name} " \
+                        f"INITCOND {self.initial_condition};"
+        return query_string
+
+    @classmethod
+    def from_yaml(cls, uda_yaml_file_path: str):
+        with Path(uda_yaml_file_path).open(mode="r", encoding="utf-8") as uda_yaml:
+            input_yaml = yaml.safe_load(uda_yaml)
+            accumulator_udf = UDFS.get((input_yaml.get("accumulator_udf_name", None)))
+            reduce_udf = UDFS.get((input_yaml.get("reduce_udf_name", None)))
+            final_udf = UDFS.get((input_yaml.get("final_udf_name", None)))
+            return UDA(
+                accumulator_udf=accumulator_udf,
+                reduce_udf=reduce_udf,
+                final_udf=final_udf,
+                **input_yaml
+            )
+
+
+def _load_all_udas():
```

Comment:
Add also type hints, so using `UDAS` will be easier.

Reply:
Fixed (also added them to UDA methods)